### ISSUE TITLE
End-to-end security

### ISSUE THREAD

**Issue by BillBennett** [OpenMobileAlliance/OMA_LwM2M_for_Developers#60]
It seems likely that proxies or PubSub will be common, in order to work around firewalls or limited reachability. In this case confidentiality and authentication of the transmitted data is lost - DTLS only protects each hop.
The PubSub draft notes that when required this security needs to be provided by the application layer.

Perhaps LWM2M should have a mechanism to negotiate this and to encrypt the body of the message? It could be left to the user, but this will probably lead to interoperability issues.

**Comment by jvermillard** [comment]
Maybe we can reuse COSE here?
https://datatracker.ietf.org/doc/draft-ietf-cose-msg/?include_text=1

**Comment by FredRodermund** [comment]
Or Object Security for CoAP (OSCOAP) which uses COSE ?
https://datatracker.ietf.org/doc/draft-selander-ace-object-security/

**Comment by hannestschofenig** [comment]
Here are my personal thoughts about this topic. 

This work is still ongoing in the IETF and recently a new working group (namely COSE) has been created. COSE aims to provide a mechanism for protecting CBOR payloads similarly to what the JOSE working group developed to protect JSON payloads. 

The PubSub draft is also still an individual working group item and will hopefully be adopted by the CORE working group. As part of the work on that document a solution for offering e2e security will have to be found. 

Finally, there is also the work in the ACE working group might provide the necessary extensions for key distribution and authorization. 

In a nutshell, it is a bit too early to tell how the final solution will look like but the COSE, ACE and CORE working groups are worthwhile to watch. 

References: 
- CBOR Object Signing and Encryption (cose) WG: https://datatracker.ietf.org/wg/cose/charter/
- Constrained RESTful Environments (core) WG: https://datatracker.ietf.org/wg/core/charter/
- Authentication and Authorization for Constrained Environments (ace) WG: https://datatracker.ietf.org/wg/ace/charter/

**Comment by hannestschofenig** [comment]
We discussed this issue at the OMA interim meeting in Paris and the participants agreed on the following way forward. 

DTLS provides channel security for protection of the communication between the LWM2M client and the LWM2M server. Additional security protection at the application layer is outside the scope of the LWM2M version 1.0. Support for this functionality will be explored in LWM2M version 1.1. We are soliciting input on the use cases.

**Comment by ThGarnier** [comment]
So this issue is (virtually) closed and is turned into a new requirement for a LWM2M next release. As said by Hannes input for use cases are welcome

**Comment by Megan-OMA** [comment]
Issue closed per Hannes' and Thierry's comments this will be addresses in release 1.1
